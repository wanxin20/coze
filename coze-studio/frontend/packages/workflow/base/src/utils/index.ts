/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export {
  SchemaExtractorParserName,
  SchemaExtractor,
  type SchemaExtractorConfig,
  type SchemaExtracted,
  type SchemaExtractorNodeConfig,
  type ParsedVariableMergeGroups,
} from './schema-extractor';

export { concatTestId, concatNodeTestId } from './concat-test-id';
export {
  type NodeResultExtracted,
  type CaseResultData,
  NodeResultExtractor,
} from './node-result-extractor';

export { parseImagesFromOutputData } from './output-image-parser';

export { reporter, captureException } from './slardar-reporter';

export { getFormValueByPathEnds } from './form-helpers';

export { isGeneralWorkflow } from './is-general-workflow';

export { isPresetStartParams, isUserInputStartParams } from './start-params';

export {
  type TraverseValue,
  type TraverseNode,
  type TraverseContext,
  type TraverseHandler,
  traverse,
} from './traverse';
export { getFileAccept } from './get-file-accept';
