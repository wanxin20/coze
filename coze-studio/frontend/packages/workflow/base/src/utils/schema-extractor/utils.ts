/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { get } from 'lodash-es';

import { type ValueExpressionDTO } from '../../types';

// Whether we upload the generated url ourselves, this is a temporary solution, after fixing the type: string - > type: image in the schema, delete this logic
export function isWorkflowImageTypeURL(str: string): boolean {
  // base64 processing
  const hostWhiteList = [
    'cC1ib3Qtd29ya2Zsb3ctc2lnbi5ieXRlZGFuY2UubmV0',
    'cC1ib3Qtd29ya2Zsb3cuYnl0ZWQub3Jn',
    'cC1ib3Qtd29ya2Zsb3cuYnl0ZWRhbmNlLm5ldA==',
    'cDI2LWJvdC13b3JrZmxvdy1zaWduLmJ5dGVpbWcuY29t',
    'cDMtYm90LXdvcmtmbG93LXNpZ24uYnl0ZWltZy5jb20=',
    'cDktYm90LXdvcmtmbG93LXNpZ24uYnl0ZWltZy5jb20=',
  ];
  const suffixWhiteList = ['image', 'jpg', 'jpeg', 'png'];

  let urlObj;

  try {
    urlObj = new URL(str);
  } catch (_) {
    return false;
  }

  const suffix =
    urlObj.searchParams?.get('x-wf-file_name')?.split('.')?.pop() ||
    urlObj.pathname.split('.').pop();

  if (!suffixWhiteList.includes(suffix)) {
    return false;
  }

  if (!hostWhiteList.includes(btoa(urlObj.hostname ?? ''))) {
    return false;
  }

  return true;
}

export const parseExpression = (expression?: ValueExpressionDTO) => {
  if (!expression) {
    return null;
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const content = get(expression, 'value.content') as any;
  if (!content) {
    return null;
  } else if (typeof content === 'string') {
    return {
      value: content,
      isImage: isWorkflowImageTypeURL(content),
    };
  } else if (content.source === 'block-output' && typeof content.name) {
    return {
      value: content.name,
      isImage: false,
    };
  } else if (
    typeof content.source === 'string' &&
    content.source.startsWith('global_variable')
  ) {
    return {
      value: content?.path?.join('.'),
      isImage: false,
    };
  }
};
