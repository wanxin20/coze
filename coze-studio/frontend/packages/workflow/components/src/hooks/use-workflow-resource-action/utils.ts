/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ProductDraftStatus,
  SchemaType,
  type FrontWorkflowInfo,
} from '@coze-workflow/base';
import { type ResourceInfo } from '@coze-arch/bot-api/plugin_develop';

import { type WorkflowResourceBizExtend } from './type';

export const parseWorkflowResourceBizExtend = (
  bizExtend?: Record<string, string>,
): WorkflowResourceBizExtend | undefined => {
  if (!bizExtend) {
    return undefined;
  }
  return {
    product_draft_status:
      bizExtend.product_draft_status !== undefined
        ? parseInt(bizExtend.product_draft_status || '0')
        : ProductDraftStatus.Default,
    external_flow_info: bizExtend.external_flow_info,
    schema_type:
      bizExtend.schema_type !== undefined
        ? parseInt(bizExtend.schema_type || '0')
        : SchemaType.DAG,
    plugin_id: bizExtend.plugin_id,
    icon_uri: bizExtend.icon_uri,
    url: bizExtend.url,
  };
};
/**
 * Convert ResourceInfo to WorkflowInfoLocal structure required to edit workflow
 * @param resource
 */
export const transformResourceToWorkflowEditInfo = (
  resource: ResourceInfo,
): Pick<
  FrontWorkflowInfo,
  | 'workflow_id'
  | 'url'
  | 'icon_uri'
  | 'name'
  | 'desc'
  | 'schema_type'
  | 'external_flow_info'
  | 'space_id'
> => {
  const bizExtend = parseWorkflowResourceBizExtend(resource.biz_extend);
  return {
    workflow_id: resource.res_id,
    url: bizExtend?.url,
    icon_uri: bizExtend?.icon_uri,
    name: resource.name,
    desc: resource.desc,
    schema_type: bizExtend?.schema_type,
    external_flow_info: bizExtend?.external_flow_info,
    space_id: resource.space_id,
  };
};
