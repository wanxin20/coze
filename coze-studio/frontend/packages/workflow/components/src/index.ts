/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @coze-arch/no-batch-import-or-export */
export { CreateWorkflowModal } from './workflow-edit';
export { FlowShortcutsHelp } from './flow-shortcuts-help';
export { WorkflowCommitList } from './workflow-commit-list';
export * from './expression-editor';
export { useWorkflowModal } from './hooks/use-workflow-modal';
export { useWorkflowList } from './hooks/use-workflow-list';
import WorkflowModalContext from './workflow-modal/workflow-modal-context';
import { type WorkflowModalContextValue } from './workflow-modal/workflow-modal-context';
import { type BotPluginWorkFlowItem } from './workflow-modal/type';
import WorkflowModal from './workflow-modal';

export { WorkflowModal, BotPluginWorkFlowItem };
export {
  useWorkflowModalParts,
  DataSourceType,
  MineActiveEnum,
  WorkflowModalFrom,
  WorkflowModalProps,
  WorkFlowModalModeProps,
  WorkflowModalState,
  WORKFLOW_LIST_STATUS_ALL,
  isSelectProjectCategory,
  WorkflowCategory,
} from './workflow-modal';

export * from './utils';
export * from './image-uploader';
export { SizeSelect, type SizeSelectProps } from './size-select';
export { Text } from './text';

export { Expression } from './expression-editor-next';
export {
  useWorkflowResourceAction,
  useWorkflowPublishEntry,
  useCreateWorkflowModal,
  useWorkflowResourceClick,
  useWorkflowResourceMenuActions,
} from './hooks/use-workflow-resource-action';
export {
  WorkflowResourceActionProps,
  WorkflowResourceActionReturn,
  WorkflowResourceBizExtend,
} from './hooks/use-workflow-resource-action/type';

export { useWorkflowProductList } from './workflow-modal/hooks/use-workflow-product-list';
export { useWorkflowAction } from './workflow-modal/hooks/use-workflow-action';
export { WorkflowModalContext, WorkflowModalContextValue };
export { useOpenWorkflowDetail } from './hooks/use-open-workflow-detail';
export { VoiceSelect } from './voice-select';
