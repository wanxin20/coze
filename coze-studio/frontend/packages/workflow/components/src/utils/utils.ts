/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import dayjs from 'dayjs';

let getIsIPadCache: boolean | undefined;
/**
 * Code provided by gpt-4
 */
export const getIsIPad = () => {
  if (typeof getIsIPadCache === 'undefined') {
    const { userAgent } = navigator;
    const isIPadDevice = /iPad/.test(userAgent);
    const isIPadOS =
      userAgent.includes('Macintosh') &&
      'ontouchstart' in document.documentElement;

    getIsIPadCache = isIPadDevice || isIPadOS;
  }

  return getIsIPadCache;
};

/* Timestamp converts text and omits the year or date*/
export const formatOmittedDateTime = time => {
  if (!time) {
    return '';
  }

  const day = dayjs(time);
  const today = dayjs();

  let formatStr: string;

  if (!today.isSame(day, 'year')) {
    // Not the year, show the year
    formatStr = 'YYYY-MM-DD HH:mm';
  } else if (!today.isSame(day, 'day')) {
    // Not the day, the display date.
    formatStr = 'MM-DD HH:mm';
  } else {
    // Show time only on the day
    formatStr = 'HH:mm';
  }

  return day.format(formatStr);
};

/** wait */
export const wait = (ms: number) =>
  new Promise(r => {
    setTimeout(r, ms);
  });

import { reporter as infraReporter } from '@coze-arch/logger';

/**
 * The slardar reporting instance used by the process
 */
export const reporter = infraReporter.createReporterWithPreset({
  namespace: 'workflow',
});
