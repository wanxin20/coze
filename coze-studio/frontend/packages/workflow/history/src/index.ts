/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { WorkflowHistoryContainerModule } from './workflow-history-container-module';
export { useClearHistory } from './hooks/use-clear-history';
export { WorkflowHistoryConfig } from './workflow-history-config';
export { createOperationReportPlugin } from './create-operation-report-plugin';

export {
  HistoryService,
  createFreeHistoryPlugin,
} from '@flowgram-adapter/free-layout-editor';
