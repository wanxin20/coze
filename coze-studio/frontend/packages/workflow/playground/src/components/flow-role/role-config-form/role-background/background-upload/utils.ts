/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type ReactCropperElement } from 'react-cropper';
import { type RefObject } from 'react';

import { floor, pick } from 'lodash-es';
import ColorThief from 'colorthief';
import { MODE_CONFIG } from '@coze-common/chat-uikit';
import { I18n } from '@coze-arch/i18n';
import { Toast } from '@coze-arch/coze-design';
import { CustomError } from '@coze-arch/bot-error';
import { type BackgroundImageInfo } from '@coze-arch/bot-api/workflow_api';
import {
  type CanvasPosition,
  type GradientPosition,
} from '@coze-arch/bot-api/developer_api';

const MIN_HEIGHT = 640;

// The function that limits the width and height of the image upload, returns true if the conditions are met, false if not
export const checkImageWidthAndHeight = (file: Blob): Promise<boolean> =>
  new Promise((resolve, reject) => {
    const fileReader = new FileReader();
    fileReader.onload = event => {
      const result = event.target?.result;
      if (!result || typeof result !== 'string') {
        reject(
          new CustomError('checkImageWidthAndHeight', 'file read invalid'),
        );
        return;
      }
      const image = new Image();
      image.src = result;
      image.onload = function () {
        if (image.height < MIN_HEIGHT) {
          Toast.error(I18n.t('bgi_upload_image_format_requirement'));
          resolve(false);
        } else if (image.complete) {
          resolve(true);
        }
      };
    };
    fileReader.onerror = () => {
      reject(new CustomError('checkImageWidthAndHeight', 'file read fail'));
    };
    fileReader.onabort = () => {
      reject(new CustomError('checkImageWidthAndHeight', 'file read abort'));
    };
    fileReader.readAsDataURL(file);
  });

export const getPictureValue = (value?: BackgroundImageInfo) => {
  if (!value) {
    return {};
  }
  return {
    url: value?.web_background_image?.origin_image_url ?? '',
    uri: value?.web_background_image?.origin_image_uri ?? '',
  };
};

export const getModeInfo = (mode: 'pc' | 'mobile') => MODE_CONFIG[mode];

// Calculate shadow position
export const computePosition = (
  mode: 'pc' | 'mobile',
  cropperRef: RefObject<ReactCropperElement>,
): GradientPosition => {
  const cropperObj = cropperRef?.current?.cropper;
  if (!cropperObj) {
    return {
      left: 0,
      right: 0,
    };
  }
  const { size } = getModeInfo(mode);
  const cropperWidth = size.width;
  const canvasData = cropperObj?.getCanvasData();
  const imgData = cropperObj?.getImageData();

  // Image left edge, distance, offset distance from the left edge of the crop area
  const imgToScreenLeft = imgData.left + canvasData?.left;
  // The distance of the picture from the right screen, > 0, the right side is not full of pictures
  const imgToScreenRight = cropperWidth - imgData.width - imgToScreenLeft;

  // Left value required for gradual change of left rendering: render left when there is a gap on the left, otherwise no need
  const leftPercent = floor(imgToScreenLeft / cropperWidth, 2);

  // The right value required for gradual change of right rendering
  const rightPercent = floor(imgToScreenRight / cropperWidth, 2);

  return {
    left: leftPercent,
    right: rightPercent,
  };
};
export const canvasPosition = (
  cropperRef: RefObject<ReactCropperElement>,
): CanvasPosition =>
  pick(cropperRef.current?.cropper.getCanvasData(), [
    'left',
    'top',
    'width',
    'height',
  ]);

export function getImageThemeColor(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const colorThief = new ColorThief();
    const img = new Image();
    img.src = url;
    img.onload = () => {
      const thiefColor = colorThief.getColor(img);
      if (thiefColor) {
        const [a, b, c] = thiefColor;
        const color = `rgba(${a}, ${b}, ${c})`;
        resolve(color);
      } else {
        reject(new CustomError('getImageThemeColor', 'not get theme color'));
      }
    };
  });
}

// Calculate theme color
export const computeThemeColor = (
  cropperRefList: RefObject<ReactCropperElement>[],
): Promise<string[]> =>
  new Promise((resolve, reject) => {
    const promises: Promise<string>[] = [];
    // Process each canvas element
    cropperRefList.forEach(cropperEle => {
      promises.push(
        new Promise<string>((resolveColor, rejectColor) => {
          const cropperObj = cropperEle.current?.cropper;
          const corp = cropperObj
            ?.getCroppedCanvas()
            ?.toDataURL('image/webp', 0.7);
          if (!corp) {
            rejectColor(
              new CustomError('computeThemeColor', 'cropper not exist'),
            );
          } else {
            getImageThemeColor(corp).then((res: string) => {
              resolveColor(res);
            });
          }
        }),
      );
    });
    Promise.all(promises)
      .then(colors => resolve(colors))
      .catch(error => reject(error));
  });
