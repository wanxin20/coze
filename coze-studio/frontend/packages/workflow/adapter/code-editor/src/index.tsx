/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { themes } from '@coze-editor/editor/preset-code';

import { convertSchema } from './utils';
import { type PreviewerProps, type EditorProps } from './interface';
import { createDarkTheme } from './components/theme';
import { Previewer } from './components/previewer';
import { BizEditor as Editor } from './components/editor';

themes.register('code-editor-dark', createDarkTheme());

export {
  Previewer,
  Editor,
  type PreviewerProps,
  type EditorProps,
  convertSchema,
};
