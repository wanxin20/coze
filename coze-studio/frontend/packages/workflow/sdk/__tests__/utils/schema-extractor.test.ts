/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { vi, describe, it, expect } from 'vitest';
import {
  type SchemaExtractorConfig,
  SchemaExtractorParserName,
  StandardNodeType,
  type WorkflowJSON,
} from '@coze-workflow/base';

import { mockCanvasSchema } from '../__mock_data__/canvas-schema';
import { schemaExtractor } from '../../src/utils';

vi.mock('lottie-web', () => ({}));

const config: SchemaExtractorConfig = {
  [StandardNodeType.Code]: [
    {
      name: 'title',
      path: 'nodeMeta.title',
    },
  ],
  [StandardNodeType.LLM]: [
    {
      name: 'title',
      path: 'nodeMeta.title',
    },
    {
      name: 'llmParam',
      path: 'inputs.llmParam',
      parser: SchemaExtractorParserName.LLM_PARAM,
    },
  ],
};

describe('schema-extractor test in @coze-workflow/sdk', () => {
  it('should extract schema result', async () => {
    const result = schemaExtractor({
      schema: mockCanvasSchema as unknown as WorkflowJSON,
      config,
    });
    await expect(result).toMatchFileSnapshot(
      './__snapshots__/schema-extractor.test.ts.snap',
    );
  });
});
