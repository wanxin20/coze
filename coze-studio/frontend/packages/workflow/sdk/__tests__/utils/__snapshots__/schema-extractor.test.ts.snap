[
  {
    "nodeId": "172934",
    "nodeType": "5",
    "properties": {
      "title": "代码",
    },
  },
  {
    "nodeId": "126762",
    "nodeType": "3",
    "properties": {
      "llmParam": {
        "prompt": "{{input}}",
        "systemPrompt": "",
      },
      "title": "大模型",
    },
  },
]